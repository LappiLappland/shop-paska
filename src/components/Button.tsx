import { ReactNode, useRef } from 'react';
import useRipple from '../hooks/useRipple';

interface ButtonProps {
  children?: ReactNode;
  className?: string;
  color?: 'primary' | 'secondary' | 'tertiary';
  rippleClassName?: string;
  type?: 'submit' | 'reset' | 'button';
  disabled?: boolean;
  dataTestid?: string;
  onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

export default function Button({
  children,
  className = '',
  color = 'primary',
  rippleClassName = '',
  type,
  dataTestid,
  disabled,
  onClick,
}: ButtonProps) {
  const buttonRef = useRef<HTMLButtonElement>(null);

  const ripple = useRipple(buttonRef, {
    color: 'bg-on-primary/focused',
    className: rippleClassName,
  });

  function getColors() {
    switch (color) {
      case 'tertiary':
        return `
        bg-tertiary text-on-tertiary
        hover:bg-tertiary/hovered
        focus-visible:bg-tertiary/focused
        active:bg-tertiary/pressed
        `;
      case 'secondary':
        return `
        bg-secondary text-on-secondary 
        hover:bg-secondary/hovered
        focus-visible:bg-secondary/focused
        active:bg-secondary/pressed
        `;
      default:
        return `
        bg-primary text-on-primary 
        hover:bg-primary/hovered
        focus-visible:bg-primary/focused
        active:bg-primary/pressed
        `;
    }
  }

  return (
    <button
      ref={buttonRef}
      data-testid={dataTestid}
      className={`
      relative flex h-10 items-center justify-center rounded-button
      px-3
      transition-colors
      ${getColors()}
      disabled:bg-on-surface/disabled-r disabled:text-on-surface
      ${className}`}
      type={type}
      disabled={disabled}
      onClick={(e) => {
        if (onClick) onClick(e);
      }}
    >
      {children}
      {ripple}
    </button>
  );
}
