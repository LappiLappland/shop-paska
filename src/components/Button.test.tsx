import { act, render } from '@testing-library/react';
import Button from './Button';

describe('Button component', () => {
  test('renders button with default primary color', () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders button with secondary color', () => {
    const { container } = render(<Button color="secondary">Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders button with tertiary color', () => {
    const { container } = render(<Button color="tertiary">Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders disabled button', () => {
    const { container } = render(<Button disabled>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls onClick handler when button is clicked', () => {
    const onClick = jest.fn();
    const { getByRole } = render(<Button onClick={onClick}>Click me</Button>);
    const button = getByRole('button');
    act(() => {
      button.click();
    })
    expect(onClick).toHaveBeenCalled();
  });
});