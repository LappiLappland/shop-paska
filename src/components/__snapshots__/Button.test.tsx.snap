// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders button with default primary color 1`] = `
<button
  class="
      relative flex h-10 items-center justify-center rounded-button
      px-3
      transition-colors
      
        bg-primary text-on-primary 
        hover:bg-primary/hovered
        focus-visible:bg-primary/focused
        active:bg-primary/pressed
        
      disabled:bg-on-surface/disabled-r disabled:text-on-surface
      "
>
  Click me
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;

exports[`Button component renders button with secondary color 1`] = `
<button
  class="
      relative flex h-10 items-center justify-center rounded-button
      px-3
      transition-colors
      
        bg-secondary text-on-secondary 
        hover:bg-secondary/hovered
        focus-visible:bg-secondary/focused
        active:bg-secondary/pressed
        
      disabled:bg-on-surface/disabled-r disabled:text-on-surface
      "
>
  Click me
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;

exports[`Button component renders button with tertiary color 1`] = `
<button
  class="
      relative flex h-10 items-center justify-center rounded-button
      px-3
      transition-colors
      
        bg-tertiary text-on-tertiary
        hover:bg-tertiary/hovered
        focus-visible:bg-tertiary/focused
        active:bg-tertiary/pressed
        
      disabled:bg-on-surface/disabled-r disabled:text-on-surface
      "
>
  Click me
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;

exports[`Button component renders disabled button 1`] = `
<button
  class="
      relative flex h-10 items-center justify-center rounded-button
      px-3
      transition-colors
      
        bg-primary text-on-primary 
        hover:bg-primary/hovered
        focus-visible:bg-primary/focused
        active:bg-primary/pressed
        
      disabled:bg-on-surface/disabled-r disabled:text-on-surface
      "
  disabled=""
>
  Click me
  <span
    class="absolute inset-0 h-full w-full overflow-hidden "
  />
</button>
`;
